src/biogeophys/SnowCoverFractionMod.F90

```diff
@@ -0,0 +1,767 @@
+    !
+    ! !ARGUMENTS:
+    class(snow_cover_fraction_type), allocatable :: scf_method  ! function result
+
+    character(len=*)        , intent(in)    :: snow_cover_fraction_method
+    type(bounds_type)       , intent(in)    :: bounds
+    type(column_type)       , intent(in)    :: col
+    type(glc_behavior_type) , intent(in)    :: glc_behavior
+    character(len=*)        , intent(in)    :: NLFilename ! Namelist filename
+    type(file_desc_t)       , intent(inout) :: params_ncid ! pio netCDF file id for parameter file
+    !
+    ! !LOCAL VARIABLES:
+
+    character(len=*), parameter :: subname = 'CreateAndInitScfMethod'
+    !-----------------------------------------------------------------------
+
+    select case (snow_cover_fraction_method)
+    case ('clm5')
+       allocate(scf_method, &
+            source = snow_cover_fraction_clm5_type())
+    case ('ny07')
+       allocate(scf_method, &
+            source = snow_cover_fraction_ny07_type())
+    case default
+       write(iulog,*) subname//' ERROR: unknown snow_cover_fraction_method: ', &
+            snow_cover_fraction_method
+       call endrun(msg = 'unknown snow_cover_fraction_method', &
+            additional_msg = errMsg(sourcefile, __LINE__))
+    end select
+
+    call scf_method%Init( &
+         bounds       = bounds, &
+         col          = col, &
+         glc_behavior = glc_behavior, &
+         NLFilename   = NLFilename, &
+         params_ncid  = params_ncid)
+
+  end function CreateAndInitScfMethod
+
+  ! ========================================================================
+  ! Methods for N&Y07 (Niu & Yang 2007) parameterization
+  ! ========================================================================
+
+  !-----------------------------------------------------------------------
+  subroutine InitNY07(this, bounds, col, glc_behavior, NLFilename, params_ncid)
+    !
+    ! !DESCRIPTION:
+    ! Initialize this instance of the NY07 parameterization
+    !
+    ! !ARGUMENTS:
+    class(snow_cover_fraction_ny07_type) , intent(inout) :: this
+    type(bounds_type)       , intent(in)    :: bounds
+    type(column_type)       , intent(in)    :: col
+    type(glc_behavior_type) , intent(in)    :: glc_behavior
+    character(len=*)        , intent(in)    :: NLFilename  ! Namelist filename
+    type(file_desc_t)       , intent(inout) :: params_ncid ! pio netCDF file id for parameter file
+    !
+    ! !LOCAL VARIABLES:
+
+    character(len=*), parameter             :: subname = 'InitNY07'
+    !-----------------------------------------------------------------------
+
+    if (use_subgrid_fluxes) then
+       write(iulog,*) 'ERROR: Attempt to use N&Y07 snow cover fraction parameterization with use_subgrid_fluxes.'
+       write(iulog,*) 'These two options are incompatible.'
+       call endrun('N&Y07 snow cover fraction parameterization incompatible with use_subgrid_fluxes')
+    end if
+
+    call this%ReadParamsNY07( &
```

Comment:
Note that I call the ReadParams method from inside the Init method (and similarly for the CLM5 parameterization), rather than from readParamsMod. This is similar to how we read namelist values for OO classes, and feels like a more modular way of dealing with parameter reads - as opposed to having a God-like readParamsMod that knows all of the parameterizations that need to read parameters. This also keeps all of the initialization of the class together in one place, rather than having it partially initialized by the Init method and partly by readParamsMod. A final consideration in designing it this way was the possibility that there could be a parameterization that doesn't need to read any parameters: If we wanted the parameter read to be handled by readParamsMod, we'd need to have a public readParams method on this class, which could be confusing if some classes don't actually need any parameters.

@ekluzek and @slevisconsulting I'd suggest that we move to a design like this moving forward for any object-oriented classes that need to read parameters.